### PR TITLE
Chore: refactor workflow with workflow engine

### PR DESCRIPTION
**Pull request by FogDong**
Signed-off-by: FogDong <dongtianxin.tx@alibaba-inc.com>

### Description of your changes

This PR deprecated all workflow and cue codes since they're removed to [workflow](https://github.com/kubevela/workflow) repository.

Instead, KubeVela now uses the workflow engine from [workflow](https://github.com/kubevela/workflow) repository to execute the workflow inside of the Application.

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: charts/vela-core/crds/core.oam.dev_applications.yaml [hunk 1]
```diff
@@ -767,16 +766,22 @@ spec:
                         description: 'UID of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids'
                         type: string
                     type: object
+                  endTime:
```
Comment (wonderflow) [inline charts/vela-core/crds/core.oam.dev_applications.yaml:769]: why we need to add a new field?

File: charts/vela-core/crds/core.oam.dev_applications.yaml [hunk 2]
```diff
@@ -860,6 +865,7 @@ spec:
                 required:
                 - finished
                 - mode
+                - status
```
Comment (wonderflow) [inline charts/vela-core/crds/core.oam.dev_applications.yaml:868]: why the newly added fields is required?

Comment (wonderflow) [reply charts/vela-core/crds/core.oam.dev_applications.yaml:868]: What about the compatiblity？

File: apis/core.oam.dev/common/types.go [hunk 3]
```diff
@@ -337,95 +312,26 @@ type PolicyStatus struct {
...
-}
-
 // WorkflowStatus record the status of workflow
 type WorkflowStatus struct {
-	AppRevision string       `json:"appRevision,omitempty"`
-	Mode        WorkflowMode `json:"mode"`
-	Message     string       `json:"message,omitempty"`
+	AppRevision string                            `json:"appRevision,omitempty"`
+	Mode        string                            `json:"mode"`
+	Phase       workflowv1alpha1.WorkflowRunPhase `json:"status"`
```
Comment (wonderflow) [inline apis/core.oam.dev/common/types.go:319]: a new field you call it `phase` in Go while call it `status` in json?

File: apis/core.oam.dev/common/types.go [hunk 4]
```diff
@@ -542,16 +422,13 @@ type ClusterPlacement struct {
 	Distribution Distribution `json:"distribution,omitempty"`
 }
 
-// ResourceCreatorRole defines the resource creator.
-type ResourceCreatorRole string
```
Comment (wonderflow) [inline apis/core.oam.dev/common/types.go:546]: why this need to be changed? Can it be removed to workflow repo?

File: go.mod [hunk 5]
```diff
@@ -51,6 +54,7 @@ require (
 	github.com/klauspost/compress v1.15.9
 	github.com/koding/websocketproxy v0.0.0-20181220232114-7ed82d81a28c
 	github.com/kubevela/prism v1.4.1-0.20220613123457-94f1190f87c2
+	github.com/kubevela/workflow v0.0.0-20220828153805-1e54f0265666
```
Comment (wonderflow) [inline go.mod:57]: we must have a concrete tag

File: pkg/apiserver/domain/service/workflow.go [hunk 6]
```diff
@@ -368,6 +369,10 @@ func (w *workflowServiceImpl) SyncWorkflowRecord(ctx context.Context) error {
 			continue
 		}
 
+		if app.Status.Workflow == nil {
+			continue
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/workflow.go:373]: why logic changed?

File: pkg/apiserver/domain/service/workflow.go [hunk 7]
```diff
@@ -414,10 +419,12 @@ func (w *workflowServiceImpl) SyncWorkflowRecord(ctx context.Context) error {
 			continue
 		}
 
-		appRevision.Spec.Application.Status.Workflow = appRevision.Status.Workflow
-		if !appRevision.Spec.Application.Status.Workflow.Finished {
-			appRevision.Spec.Application.Status.Workflow.Finished = true
-			appRevision.Spec.Application.Status.Workflow.Terminated = true
+		if appRevision.Status.Workflow != nil {
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/workflow.go:422]: already checked in L372?

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 8]
```diff
@@ -211,53 +217,45 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
-	case common.WorkflowStateFinished:
-		logCtx.Info("Workflow state=Finished")
-		if status := app.Status.Workflow; status != nil && status.Terminated {
-			return r.result(nil).ret()
+			if result, err := r.gcResourceTrackers(logCtx, handler, common.ApplicationRunning, false, isUpdate); err != nil {
+				return result, err
+			}
 		}
-	case common.WorkflowStateSkipping:
+	case workflowv1alpha1.WorkflowRunSkipped:
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:255]: we shouldn't use `WorkflowRun` as a prefix here. We need a common state such as "WorkflowSkipped"

Comment (wonderflow) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:255]: the same on above

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 9]
```diff
@@ -282,7 +280,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		Reason:             condition.ReasonReconcileSuccess,
 	})
 	r.Recorder.Event(app, event.Normal(velatypes.ReasonDeployed, velatypes.MessageDeployed))
-	return r.gcResourceTrackers(logCtx, handler, phase, true, true)
+	return r.gcResourceTrackers(logCtx, handler, phase, true, false)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:283]: why true -> false?

will it incompatible?

File: apis/core.oam.dev/common/types.go (hunk 3 above)
Comment (FogDong) [reply apis/core.oam.dev/common/types.go:319]: That's what we do in application status...

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go [hunk 10]
```diff
@@ -4044,62 +4076,6 @@ var _ = Describe("Test Application Controller", func() {
 		}))
 	})
 
-	It("app record execution state with controllerRevision", func() {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go:4047]: is this test case deleted?

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go [hunk 11]
```diff
@@ -18,41 +18,57 @@ package application
...
 	oamProvider "github.com/oam-dev/kubevela/pkg/workflow/providers/oam"
 	terraformProvider "github.com/oam-dev/kubevela/pkg/workflow/providers/terraform"
-	"github.com/oam-dev/kubevela/pkg/workflow/tasks"
-	wfTypes "github.com/oam-dev/kubevela/pkg/workflow/types"
+	"github.com/oam-dev/kubevela/pkg/workflow/template"
 )
 
+func init() {
+	if err := stdlib.SetupBuiltinImports(); err != nil {
+		klog.ErrorS(err, "Unable to set up builtin imports")
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/generator.go:67]:
```suggestion
		klog.ErrorS(err, "Unable to set up builtin imports on package initialization")
```

File: apis/core.oam.dev/common/types.go (hunk 4 above)
Comment (FogDong) [reply apis/core.oam.dev/common/types.go:546]: We can use string directly instead.

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go [hunk 12]
```diff
@@ -83,84 +104,120 @@ func (h *AppHandler) GenerateApplicationSteps(ctx monitorContext.Context,
...
+func generateWorkflowRun(af *appfile.Appfile, app *v1beta1.Application, appRev string) (*workflowv1alpha1.WorkflowRun, error) {
+	revAndSpecHash, err := workflow.ComputeWorkflowRevisionHash(appRev, app)
+	if err != nil {
+		return nil, err
+	}
+	anno := make(map[string]string)
+	if af.Debug {
+		anno[wfTypes.AnnotationWorkflowRunDebug] = "true"
+	}
+	run := &workflowv1alpha1.WorkflowRun{
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/generator.go:146]: why we need to leverage the workflowRun CRD?

File: pkg/apiserver/domain/service/workflow.go (hunk 7 above)
Comment (FogDong) [reply pkg/apiserver/domain/service/workflow.go:422]: Not the same.

File: pkg/cue/README.md [hunk 13]
```diff
@@ -1,10 +0,0 @@
-# Develop Tips
-
-The following packages need to be tested without external/tool dependencies, So that the cue-sh/unity can do verification
```
Comment (wonderflow) [inline pkg/cue/README.md:3]: we do need to make such kinds of packaged be tested included by CUE  community.

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go (hunk 9 above)
Comment (FogDong) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:283]: Just change the variable from `isPatch` to `isUpdate`  to make it more easy to understand.

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go (hunk 10 above)
Comment (FogDong) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go:4047]: Yes, application revision will replace controller revision and the logic of controller revision is deleted.

File: charts/vela-core/crds/core.oam.dev_applications.yaml (hunk 1 above)
Comment (FogDong) [reply charts/vela-core/crds/core.oam.dev_applications.yaml:769]: To store the end time of workflow.

File: charts/vela-core/crds/core.oam.dev_applications.yaml (hunk 2 above)
Comment (FogDong) [reply charts/vela-core/crds/core.oam.dev_applications.yaml:868]: Will make it a omit field. This field shows the status of workflow.

File: go.mod (hunk 5 above)
Comment (FogDong) [reply go.mod:57]: We can use git commit tag first and change to a concrete tag before we release 1.6.

File: pkg/apiserver/domain/service/workflow.go (hunk 6 above)
Comment (FogDong) [reply pkg/apiserver/domain/service/workflow.go:373]: `app.Status.Workflow` is a pointer, we did not check the pointer before we refer it before.

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go (hunk 12 above)
Comment (FogDong) [reply pkg/controller/core.oam.dev/v1alpha2/application/generator.go:146]: We're not using the CRD, just the workflowRun struct to use to workflow engine.

File: pkg/cue/README.md (hunk 13 above)
Comment (FogDong) [reply pkg/cue/README.md:3]: This is removed to workflow repo.

File: apis/core.oam.dev/common/types.go (hunk 4 above)
Comment (wonderflow) [reply apis/core.oam.dev/common/types.go:546]: the string type lose the constraint

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 14]
```diff
@@ -211,53 +218,45 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	handler.addAppliedResource(true, app.Status.AppliedResources...)
 	app.Status.AppliedResources = handler.appliedResources
 	app.Status.Services = handler.services
+	isUpdate := app.Status.Workflow.Message != "" && run.Status.Message == ""
+	run.Status.Phase = workflowState
+	app.Status.Workflow = workflow.ConvertWorkflowStatus(run.Status, app.Status.Workflow.AppRevision)
 	switch workflowState {
-	case common.WorkflowStateInitializing:
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:215]: it‘s ommitted?

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 15]
```diff
@@ -211,53 +218,45 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
-			return r.endWithNegativeCondition(logCtx, app, condition.ErrorCondition(common.WorkflowCondition.String(), errors.WithMessage(err, "DoWorkflowFinish")), common.ApplicationRunningWorkflow)
-		}
+		r.doWorkflowFinish(app, workflowState)
 		app.Status.SetConditions(condition.ReadyCondition(common.WorkflowCondition.String()))
 		r.Recorder.Event(app, event.Normal(velatypes.ReasonApplied, velatypes.MessageWorkflowFinished))
 		logCtx.Info("Application manifests has applied by workflow successfully")
 		if !EnableReconcileLoopReduction {
-			return r.gcResourceTrackers(logCtx, handler, common.ApplicationWorkflowFinished, false, true)
-		}
-	case common.WorkflowStateFinished:
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:253]: why this case removed？

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go [hunk 16]
```diff
@@ -83,84 +104,120 @@ func (h *AppHandler) GenerateApplicationSteps(ctx monitorContext.Context,
...
-		if err != nil {
-			return nil, err
-		}
-		tasks = append(tasks, task)
+	run, err := generateWorkflowRun(af, app, appRev.Name)
+	if err != nil {
+		return nil, nil, err
 	}
-	return tasks, nil
+	executor.InitializeWorkflowRun(run)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/generator.go:113]: 使用CR作为函数输入容易掉进一个常见的坑，就是在函数里认为就是在操作CR，无意中加入 get CR，update status 的逻辑。即使现在没有，也很容易让后来者掉坑。

防止这种情况最好的办法就是不用CR，可以复用 spec 的结构体，spec、status 作为2个输入。

**Review by wonderflow — APPROVED** [review]
generally LGTM

File: apis/core.oam.dev/common/types.go (hunk 4 above)
Comment (FogDong) [reply apis/core.oam.dev/common/types.go:546]: If we want the constraint, we need to add the constraint to the workflow repo but I don't think it's appropriate.

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go (hunk 14 above)
Comment (FogDong) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:215]: Yes.

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go (hunk 15 above)
Comment (FogDong) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:253]: There's no redundant finished state in workflow anymore.

File: apis/core.oam.dev/common/types.go (hunk 4 above)
Comment (wonderflow) [reply apis/core.oam.dev/common/types.go:546]: ok

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go [hunk 17]
```diff
@@ -64,15 +80,20 @@ func (h *AppHandler) GenerateApplicationSteps(ctx monitorContext.Context,
 	app *v1beta1.Application,
 	appParser *appfile.Parser,
 	af *appfile.Appfile,
-	appRev *v1beta1.ApplicationRevision) ([]wfTypes.TaskRunner, error) {
+	appRev *v1beta1.ApplicationRevision) (*workflowv1alpha1.WorkflowRun, []wfTypes.TaskRunner, error) {
 
 	handlerProviders := providers.NewProviders()
-	kube.Install(handlerProviders, app, h.r.Client, h.Dispatch, h.Delete)
+	kube.Install(handlerProviders, h.r.Client, auth.GetUserInfoInAnnotation(&app.ObjectMeta),
```
Comment (Somefive) [inline pkg/controller/core.oam.dev/v1alpha2/application/generator.go:86]: Mark: delete auth.GetUserInfoInAnnotation(&app.ObjectMeta)

**Review by Somefive — APPROVED** [review]

**Review by barnettZQG — APPROVED** [review]
lgtm

**Pull request merged**